src/common/JSHandle.ts

```diff
@@ -212,64 +182,57 @@ export class JSHandle<T = unknown> {
    * const children = [];
    * for (const property of properties.values()) {
    *   const element = property.asElement();
-   *   if (element)
+   *   if (element) {
    *     children.push(element);
+   *   }
    * }
    * children; // holds elementHandles to all children of document.body
    * ```
    */
   async getProperties(): Promise<Map<string, JSHandle>> {
-    assert(this.#remoteObject.objectId);
-    const response = await this.#client.send('Runtime.getProperties', {
-      objectId: this.#remoteObject.objectId,
-      ownProperties: true,
+    const keys = await this.evaluate(object => {
+      return Object.keys(object);
     });
-    const result = new Map<string, JSHandle>();
-    for (const property of response.result) {
-      if (!property.enumerable || !property.value) {
-        continue;
-      }
-      result.set(property.name, createJSHandle(this.#context, property.value));
-    }
-    return result;
+    const entries = await Promise.all(
+      keys.map(async key => {
+        const handle = await this.getProperty(key);
+        return [key, handle] as const;
+      })
+    );
+    return new Map(entries);
   }
 
   /**
-   * @returns Returns a JSON representation of the object.If the object has a
-   * `toJSON` function, it will not be called.
-   * @remarks
+   * @returns A vanilla object representing the serializable portions of the
+   * referenced object.
+   * @throws Throws if the object cannot be serialized due to circularity.
    *
-   * The JSON is generated by running {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify | JSON.stringify}
-   * on the object in page and consequent {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse | JSON.parse} in puppeteer.
-   * **NOTE** The method throws if the referenced object is not stringifiable.
+   * @remarks
+   * If the object has a `toJSON` function, it *will not* be called.
    */
-  async jsonValue<T = unknown>(): Promise<T> {
-    if (this.#remoteObject.objectId) {
-      const response = await this.#client.send('Runtime.callFunctionOn', {
-        functionDeclaration: 'function() { return this; }',
-        objectId: this.#remoteObject.objectId,
-        returnByValue: true,
-        awaitPromise: true,
-      });
-      return valueFromRemoteObject(response.result) as T;
+  async jsonValue(): Promise<T> {
+    if (!this.#remoteObject.objectId) {
+      return valueFromRemoteObject(this.#remoteObject);
+    }
+    const value = await this.evaluate(object => {
```

Comment:
Is it equivalent? I was under the impression that evaluate might not return by value. That means you might get a reference to an object instead of JSON?

Reply:
This is equivalent. Check tests.